earth_enterprise/src/fusion/autoingest/sysman/StateUpdater.cpp

```diff
@@ -278,16 +291,76 @@ class StateUpdater::SetStateVisitor : public default_dfs_visitor {
     }
 };
 
-void StateUpdater::SetStateForRefAndDependents(
+
+class StateUpdater::SetBlockingStateVisitor : public StateUpdater::VisitorBase {
+  private:
+    using BlockedSet = std::unordered_set<SharedString>;
+
+    // This is a shared_ptr because the visitor will be copied several times
+    // during the course of the traversal.
+    const std::shared_ptr<BlockedSet> hasBlockingInputs;
+    const std::shared_ptr<BlockedSet> hasBlockingChildren;
+
+    void MarkParentsForLaterProcessing(AssetHandle<AssetVersionImpl> & version) const override {
+      if (AssetDefs::IsBlocking(newState)){
+        hasBlockingChildren->insert(version->parents.begin(), version->parents.end());
+        hasBlockingInputs->insert(version->listeners.begin(), version->listeners.end());
+      }
+    }
+
+  public:
+    SetBlockingStateVisitor(StateUpdater & updater, DependentStateTree & tree, AssetDefs::State newState) :
+        VisitorBase(updater, tree, newState),
+        hasBlockingInputs(std::make_shared<BlockedSet>()),
+        hasBlockingChildren(std::make_shared<BlockedSet>()) {}
+
+    // This function is called after the DFS has completed for every vertex
+    // below this one in the tree. Thus, we don't calculate the state for an
+    // asset until we've calculated the state for all of its inputs and
+    // children.
+    void finish_vertex(
+        DependentStateTreeVertexDescriptor vertex,
+        const DependentStateTree &) const {
+      const AssetVertex & data = tree[vertex];
+      notify(NFY_PROGRESS, "Calculating state for '%s'", data.name.toString().c_str());
+
+      // Set the state for assets in the dependent tree.
+      if (data.inDepTree) {
+        SetState(vertex, newState, {0, 0}, true);
+      }
+      else {
+        // Set the state to blocked if needed, otherwise do nothing
+        auto version = updater.storageManager->Get(data.name);
+        if (hasBlockingInputs->find(data.name) != hasBlockingInputs->end() && 
+            version->InputStatesAffectMyState(AssetDefs::Blocked, true)) {
+          SetState(vertex, AssetDefs::Blocked, {0,0}, true);
```

Comment:
Minor: after this point you can remove `data.name` from `hasBlockingInputs`. I don't know if that will make it faster, though - it's extra work at this point to remove the name, but it's fewer names to search through for future vertices. Same applies to the `hasBlockingChildren` statement.

Reply:
I actually had that initially, but I took it out, since `erase` would have to iterate through the list to find it. I wasn't sure if that would be worth the trade-off. But it looks like maybe I can `erase` using the iterator returned from `find` to do it much faster.

Reply:
Yeah, I agree. I don't know if it will make things faster or not. It's up to you if you want to implement it.

Reply:
1. Since the documentation for `erase` says that the version that takes a position is constant time, I figured it was probably a little faster to do that.
2. It's now always checking to see if `data.name` is in `hasBlockingChildren` and `hasBlockingInputs`. The rationale is that, since I'm removing `data.name` from the sets, I should check both since it could theoretically be in both (I've no idea if that's possible in actuality). But beyond that, I think it makes the logic a bit cleaner, and I suspect it makes the unit test coverage a little better since there are fewer "hidden" branches. Also, the impact of sometimes calling `SetState` a second time should be minimal since it will return right away if the state isn't changing. It's probably a bit of a wash on the whole, performance-wise. But this seems good to me. 
As always, open to discussion.